flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp

```diff
@@ -144,12 +145,340 @@ mlir::Value calculateTripCount(fir::FirOpBuilder &builder, mlir::Location loc,
+///   fir.store %1 to %i_decl#1 : !fir.ref<i32>
+///   ...
+/// }
+/// ```
+/// this function would return the first 2 ops in the `fir.do_loop`'s region.
+llvm::SetVector<mlir::Operation *>
+extractIndVarUpdateOps(fir::DoLoopOp doLoop) {
+  mlir::Value indVar = doLoop.getInductionVar();
+
+  // The list of values related to the induction variable. These could be the IV
+  // itself of the results of ops where the IV is an operand.
+  llvm::DenseSet<mlir::Value> indVarRelatedVals;
+  // We start with the IV as the only item in the list.
+  indVarRelatedVals.insert(indVar);
+  llvm::SetVector<mlir::Operation *> indVarUpdateOps;
+
+  for (mlir::Operation &op : doLoop.getRegion().getOps()) {
+    auto opTakesValueAsOperand = [&](mlir::Value value) {
+      return llvm::any_of(op.getOperands(), [&](mlir::Value operand) {
+        return operand == value;
+      });
+    };
+
+    if (llvm::any_of(indVarRelatedVals, [&](mlir::Value value) {
+          return opTakesValueAsOperand(value);
+        })) {
+      indVarUpdateOps.insert(&op);
+
+      for (mlir::Value result : op.getResults())
+        indVarRelatedVals.insert(result);
+    }
+  }
+
+  return std::move(indVarUpdateOps);
+}
+
+/// Starting with a value and the end of a defintion/conversion chain, walk the
+/// chain backwards and collect all the visited ops along the way. For example,
+/// given this IR:
+/// ```
+/// %c10 = arith.constant 10 : i32
+/// %10 = fir.convert %c10 : (i32) -> index
+/// ```
+/// and giving `%10` as the starting input: `link`, `defChain` would contain
+/// both of the above ops.
+mlir::LogicalResult
+extractDefChain(mlir::Operation *link,
+                llvm::SmallVectorImpl<mlir::Operation *> &defChain) {
+  if (mlir::isa_and_present<mlir::arith::ConstantOp>(link)) {
+    defChain.push_back(link);
+    // A `ConstantOp` marks the beginning of the chain, so we can stop here.
+    return mlir::success();
+  }
+
+  if (auto convertOp = mlir::dyn_cast_if_present<fir::ConvertOp>(link)) {
+    // For a `ConvertOp` we still have links up the chain to walk, so
+    // recursively do so.
+    mlir::LogicalResult result =
+        extractDefChain(convertOp.getValue().getDefiningOp(), defChain);
+
+    if (failed(result))
+      return result;
+
+    defChain.push_back(link);
+    return mlir::success();
+  }
+
+  std::string opStr;
+  llvm::raw_string_ostream opOs(opStr);
+  opOs << "Unexpected operation: " << *link;
+  return mlir::emitError(link->getLoc(), opOs.str());
+}
+
+/// Starting with `outerLoop` collect a perfectly nested loop nest, if any. This
+/// function collects as much as possible loops in the nest; it case it fails to
+/// recognize a certain nested loop as part of the nest it just returns the
+/// parent loops it discovered before.
+mlir::LogicalResult
+collectLoopNest(fir::DoLoopOp outerLoop,
+                const llvm::SmallVectorImpl<mlir::Value> &outerLoopLiveIns,
+                LoopNestToIndVarMap &loopNest) {
+  loopNest.try_emplace(
```

Comment:
Doesn't the outermost also need to be unordered to be parallelized?

Reply:
That's a pre-condition to apply `DoConcurrentConversion` rewriter pattern: it will never applied to a `do concurrent` loop unless it is `unordered` since this what we mark as illegal ops ([see call to target.addDynamicallyLegalOp below](https://github.com/ROCm/llvm-project/blob/amd-trunk-dev/flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp#L598)).